Metrics collection profiles (#1298)
openshift/enhancements: enhancements/monitoring/metrics-collection-profiles.md

Users currently can not control the ServiceMonitors scraped by Prometheus since
some of the metrics collected are essential for other parts of the system to
function properly: alerting rules, console dashboards, resource metrics API,
horizontal/vertical pod autoscaling and Red Hat Telemetry. Users also are not
allowed to tune the interval at which Prometheus scrapes targets as this again
can have unforeseen results that can hinder the platform: a low scrape interval
value may overwhelm the platform Prometheus instance while a high interval value
may render some of the default alerts ineffective.

The goal of this proposal is to allow users to pick their desired level of
scraping while limiting the impact this might have on the platform, via
resources under the control of the cluster-monitoring-operator and other
platform operators.

Furthermore to assess the viability of the metrics collection profile feature
the monitoring team performed a detailed analysis of its impact in an OpenShift
cluster. The analysis consisted of a test where an OpenShift cluster would run
three replicas of the OpenShift Prometheus instance but each replica would be

> configured to a different metrics collection profile (`full`, `minimal`). Then

Reviewer: This mentions three replicas but only two profiles used. Could this be more consistent?

	// profile, Prometheus only collects metrics necessary for the default
	// platform alerts, recording rules, telemetry and console dashboards.
	CollectionProfile CollectionProfile `json:"collectionProfile,omitempty"`
}
```

### Implementation Details/Notes/Constraints

Each OpenShift team that wants to adopt this feature will be responsible for
providing the different monitors. This work is not trivial. Dependencies between
operators and their metrics exist. This makes it difficult for developers to
determine whether a given metric can be excluded from the `minimal` profile or
not. To aid teams with this effort the monitoring team will provide:
- a CLI tool that consumes a ServiceMonitor and a list of metrics and
  generates the `minimal` ServiceMonitor for that operator.
- an origin/CI test that validates for all Alerts and PrometheusRules that the
  metrics used by them are present in the `keep` expression of the
  ServiceMonitor for the `minimal` profile

### Risks and Mitigations

- How are monitors supposed to be kept up to date? In 4.12 a metric that wasn't
  being used in an alert is now required, how does the monitor responsible for
  that metric gets updated?

>   - The origin/CI test mentioned in the previous section will fail if ther is a
>     resouce (Alerts/PrometheusRules/Dashboards) using a metric which is not

Reviewer: ```suggestion
  - The origin/CI test mentioned in the previous section will fail if there is a
    resource (Alerts/PrometheusRules/Dashboards) using a metric which is not
```

    present in the ServiceMonitor;

- A new profile is added, what will happen to operators that had implemented
  metrics collection profiles but did not implement the latest profile.
  - The monitoring team will use the list under 
  [Infrastruture needed](#infrastructure-needed-optional) to help the developers
  with the addoption of the new metrics collection profile.
  - Furthermore, we plan on adding extra origin/CI tests to validate each new
    profile if applicable.

- What happens if CMO pick-up an unsupported collection profile value?

>   - Report Degraded=False and fail reconciliation.

Reviewer: Can this  bit be made a bit more precise? What exactly is the error scenario?

### Drawbacks

- Extra CI cycles

## Design Details

### Open Questions

> - Should we add future profiles? 

Reviewer: I think another open question would be, if we'd add another profile, how would we validate it?
Our current validation strategy consists of comparing metrics we use (In Rules and such) against regexes in relabel keep actions. 
For hypothetical future profiles this strategy is unclear. Since the process of adding a new profile is also a little hand-wavy (keep a list of teams to follow up with) maybe it could be an option to declare adding another profile outside of this particular proposal?
This feature is valuable as is (with only the full and minimal profile) and its unclear whether we'll add a third, so maybe we should exclude this scenario explicitly?

Author: What if we move this to `Risks and Mitigations` and detail some of the topics you highlighted there? I'm also happy with just moving it to a `Non goal` but I believe having it detailed in `Risks and Mitigations` might bring more value in the long run. WDYT?

Author: I did as I suggested in last commit, if you feel I didn't hit the mark I'm happy with moving it to the `Non goals`

Reviewer 2: (+1 on moving profile additions outside of this proposal) 🙂

- Unit tests in CMO to validate that the correct monitors are being selected
- E2E tests in CMO to validate that everything works correctly 

> - For the `minimal` profile, origin/CI test to validate that every metrics used
> in a resource (Alerts/PrometheusRules/Dashboards) exist in the `keep` expresion
> of a

Reviewer: ```suggestion
- For the `minimal` profile, origin/CI test to validate that every metric used
in a resource (Alerts/PrometheusRules/Dashboards) exist in the `keep` expression
of a minimal ServiceMonitor.
```

---
title: metrics-collection-profiles
authors:
  - @JoaoBraveCoding
reviewers:
  - @openshift/openshift-team-monitoring
approvers:
  - TBD
api-approvers: "None"
creation-date: 2022-12-06
last-updated: 2023-07-19
tracking-link:
  - https://issues.redhat.com/browse/MON-2483
  - https://issues.redhat.com/browse/MON-3043
---

# Metrics collection profiles

## Summary

The core OpenShift components ship a large number of metrics. A 4.12-nightly
cluster on AWS (3 control plane nodes + 3 worker nodes) currently produces
around 350,000 unique timeseries, and adding optional operators increases that
number. Users have repeatedly asked for a supported method of making Prometheus
consume less memory and CPU, either by increasing the scraping interval or by
scraping fewer targets.

These modifications are currently not possible, because the OpenShift operators

> manage their service/pod monitors (either directly or via the cluster version

Author: I added a section for terms to avoid always having to name the 3 CRDs

operator) and any manual modification would be reverted. This proposal outlines
a solution for allowing users to control the amount of data being collected
which can meet their needs.

## Motivation

OpenShift is an opinionated platform, and the default set of metrics has of
course been crafted to match what we think the majority of users might need.
Nevertheless, users have repeatedly asked for the ability to reduce the amount
of memory consumed by Prometheus either by lowering the Prometheus scrape

> intervals or by modifying ServiceMonitors.

Author: I added a section for terms to avoid always having to name the 3 CRDs

> Users currently can not control the ServiceMonitors scraped by Prometheus since

Author: I added a section for terms to avoid always having to name the 3 CRDs

some of the metrics collected are essential for other parts of the system to
function properly: alerting rules, console dashboards, resource metrics API,
horizontal/vertical pod autoscaling and Red Hat Telemetry. Users also are not
allowed to tune the interval at which Prometheus scrapes targets as this again
can have unforeseen results that can hinder the platform: a low scrape interval
value may overwhelm the platform Prometheus instance while a high interval value
may render some of the default alerts ineffective.

The goal of this proposal is to allow users to pick their desired level of
scraping while limiting the impact this might have on the platform, via
resources under the control of the cluster-monitoring-operator and other

> platform operators.

Author: Since we now only support two profiles and it's out of the goal of adding further profiles I feel like this is not relevant anymore. Since if a developer does not adhere to this feature their operator will just default to exporting all metrics.

for OpenShift 4.11 the 5 components most often reported as the biggest producers
are: the Kubernetes API servers, the Kubernetes schedulers, kube-state-metrics,
kubelet and the network daemon.

### User Stories

- As an OpenShift cluster administrator, I want to lower the amount of resources
  consumed by Prometheus in a supported way, so I can choose between different
  metrics collection profiles, e.g `full` or `minimal`.
- As an OpenShift developer, I want a supported way to collect a subset of the
  metrics exported by my operator and operands.

### Goals

- Give users a supported and documented method to pick the amount of metrics
  being collected by the platform, in a way that allows
  cluster-monitoring-operator and other operators to provide their own
  defaulting.
- Other components of the cluster (console, HPA, alerts) should not be
  negatively affected by this change.
- Openshift developers may optionally adhere to this change, developers that do
  not adhere should not be impacted.

### Non-Goals

- Dynamically adjusting metrics scraped at runtime based on heuristics.
- Allowing users to explicitly set a list or a regex of metrics that they would

> want to be included in a given profile.

Author: I added a section for terms to avoid these confusions. Here I meant OpenShift cluster-admins  i.e. end users

## Proposal

This enhancement leverages label selectors to allow cluster-monitoring-operator
to select the appropriate monitors ([Pod|Service]Monitor) for a given metrics

> collection profile. 

Author: I added a section for terms to avoid always having to name the 3 CRDs

Given this, the following steps would be necessary to implement this
enhancement:

- expose a configuration option in the cluster-monitoring-operator ConfigMap
  that allows selecting the metrics collection profile;
- OCP components that would like to implement metrics collection profiles would
  have to provide X monitors. Where X is the number of metrics collection
  profiles that we would support.

### Workflow Description

The cluster-monitoring-operator (CMO) will be extended with a new field,

> `collectionProfile` under `prometheusK8s`. This new field will then allow users

Author: I added a section for terms to avoid these confusions. Here I meant OpenShift cluster-admins i.e. end users

to configure their desired metrics collection profile while preserving the
platform functionality (console, HPA and alerts).

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      collectionProfile: full 
```

The different profile names would be pre-defined by the OpenShift monitoring
team. Once a profile is selected CMO then updates the platform Prometheus CR to
select pod and service monitors as the union of the 2 sets:
- pod and service monitors with the 
  `monitoring.openshift.io/collection-profile: <selected profile>` label.
- pod and service monitors without the
  `monitoring.openshift.io/collection-profile` profile label present, to retain

>   the default behaviour (for components that choose to not opt-in).

Author: I added a section for terms to avoid always having to name the 3 CRDs

OpenShift teams can decide if they wanted to adopt this feature. Without any
change to a ServiceMonitor, if a user picks a profile in the CMO config, things

> should work as they did before. When an OpenShift team wants to implement

Author: I added a section for terms to avoid always having to name the 3 CRDs

    interval: 30s
    port: https
    scheme: https
    tlsConfig:
      <...>
  jobLabel: k8s-app
  selector:
    matchLabels:
      k8s-app: telemeter-client
---
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    k8s-app: telemeter-client
    monitoring.openshift.io/collection-profile: minimal
  name: telemeter-client-minimal
  namespace: openshift-monitoring
spec:
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    interval: 30s
    port: https
    scheme: https
    tlsConfig:
      <...>
  jobLabel: k8s-app
  selector:
    matchLabels:
      k8s-app: telemeter-client
  metricRelabelings:
  - sourceLabels: [__name__]
    action: keep
    regex: "federate_samples|federate_filtered_samples"
 ```

Note: 
- the `metricRelabeling` section only keeping two metrics, while the rest is
dropped.
- the metrics in the `keep` section were obtained with the help of a script that
  parsed all Alerts, PrometheusRules and Console dashboards to determine what
  metrics were actually being used.

Finally, a team that adopts the metrics collection profile feature should also 
add themselves and their component to the list under 

> [Infrastruture needed](#infrastructure-needed-optional)

Author: I believe so, in case this is not convenient I would expect the official reviewers to pipe up

#### Variation [optional]

NA

### API Extensions

- Addition of a new field to the `cluster-monitoring-config` ConfigMap

```go
type PrometheusK8sConfig struct {
	// Defines the metrics collection profile that Prometheus uses to collect
	// metrics from the platform components. Supported values are `full` or
	// `minimal`. In the `full` profile (default), Prometheus collects all
	// metrics that are exposed by the platform components. In the `minimal`
	// profile, Prometheus only collects metrics necessary for the default
	// platform alerts, recording rules, telemetry and console dashboards.
	CollectionProfile CollectionProfile `json:"collectionProfile,omitempty"`
}
```

### Implementation Details/Notes/Constraints

Each OpenShift team that wants to adopt this feature will be responsible for
providing the different monitors. This work is not trivial. Dependencies between
operators and their metrics exist. This makes it difficult for developers to
determine whether a given metric can be excluded from the `minimal` profile or
not. To aid teams with this effort the monitoring team will provide:

> - a CLI tool that consumes a ServiceMonitor and a list of metrics and
>   generates the `minimal` ServiceMonitor for that operator.

Reviewer 2: ```suggestion
- a CLI tool that consumes a rule file and a list of metrics and
  generates the `minimal``metricRelabelings` for that operator.
```
(cc @jan--f)

Author: I'm confused about this comment, this bullet point refers to https://github.com/openshift/cluster-monitoring-operator/pull/2029 which the goal was always to provide a CLI tool to help the generation of ServiceMonitors

Reviewer 2: Ah sorry, let me loop you in on this. In the latest biweekly 1-1 with @jan--f, I talked a bit about how working with `ServiceMonitor` may open us up to be responsible for fields within that schema that are not immediately related to the effort. A better approach would be to take in the rule file that the team relies on, and generate a `metricRelabelings` schema, which they can put in their `ServiceMonitor`, `PodMonitor`, or `Probe`, without us having to worry about the different kinds.

Author: Interesting, do have an example of which fields we could end up being responsible for? I'm not 100% sold on the idea since this would imply more manual actions for the OCP teams that would want to implement this feature. They would have to run this feature Copy&Paste and add an annotation. Furthermore, I'm not too concerned about the different kinds since AFAIK most projects just use `ServiceMonitor`. 
Regardless if this is really the direction we want I'm happy to comply.

Reviewer: We discussed two things:

1. The tool adding the needed relabelings should not generate the full ServiceMonitor with all fields that are essentially unrelated to the collection profiles. So it could either produce only the relabel configs or (which Pranshuand I didn't discuss at the time) modify a given ServiceMonitor and add the needed relabelings. The latter might be more complicated since we'd have to pay attention to preexisting relabel configs, since they interact with each other and order is important. The idea was to have to tool only assemble the keep relabel action and leave it to developers to integrate that into their ServiceMonitor.
2. For ease of use it would be nice if this tool knew how to parse Prometheus rule files (or PrometheusRule CRs) and maybe also Grafana dashboards? This is probably easier to add to a script then to document for developers how to extract the needed metrics. A script could also query a running Prometheus instance for a lost of rules and extract metric names from that.

Author: @jan--f @rexagod https://github.com/openshift/cluster-monitoring-operator/pull/2053 let me know if this is what you had in mind. This basically works like so. 
```bash
→ go run hack/monitorgen/main.go --minimal-metrics-file=metrics.txt 
metricRelabelings:
- action: keep
  regex: (metric2|metric3|metric4)
  sourceLabels:
  - __name__
```
I still feel like parsing the SM and generating the full ServiceMonitor it's a friendlier workflow but I might be missing some context of the discussions

Reviewer: I agree that generating a full SM is friendlier. However we'd have to address other existing relabel configs, besides existing drop actions (which might incidentally be totally valid to keep). E.g. an already existing keep action will drop everything not explicitly kept, so two keep actions will only retain the intersection.
Basically things are more complicated then telling users "give us a service monitor and we just add whats needed". By just providing our relabel config with the message "this is what we need, please integrate that" might be the clearer message.

Author: I see, yes that certainly makes sense 👍 but then https://github.com/openshift/cluster-monitoring-operator/pull/2053 works right?

Reviewer 3: My 2 cents: I'm not sure that we already have monitors using the keep action (on the other side, we have some drop usage). The tool could be "smart" enough to flag the problem if it arises though.
I feel that providing the transformed monitors is the best approach because people aren't familiar with relabeling configs.

Reviewer: Sure, producing a best guess modified SM is also an option and probably just works for the majority of users (since they don't have any preexisting relabel configs). We should however be cautious when there are existing relabel configs and flag the produced output for human verification.

- an origin/CI test that validates for all Alerts and PrometheusRules that the
  metrics used by them are present in the `keep` expression of the

>   ServiceMonitor for the `minimal` profile
> 

Reviewer 2: ```suggestion
  respective CRD for the `minimal` profile.
- an origin/CI test that validates that none of the critical metrics are being `drop`ped, even in a `minimal` configuration.
```
(cc @jan--f)

### Risks and Mitigations

- How are monitors supposed to be kept up to date? In 4.12 a metric that wasn't
  being used in an alert is now required, how does the monitor responsible for
  that metric gets updated?
  - The origin/CI test mentioned in the previous section will fail if there is a
    resource (Alerts/PrometheusRules/Dashboards) using a metric which is not

>     present in the ServiceMonitor;

Reviewer 2: ```suggestion
    present in the CRD in question;
```

- A new profile is added, what will happen to operators that had implemented
  metrics collection profiles but did not implement the latest profile.
  - The monitoring team will use the list under 
  [Infrastruture needed](#infrastructure-needed-optional) to help the developers
  with the addoption of the new metrics collection profile.
  - Furthermore, we plan on adding extra origin/CI tests to validate each new

>     profile if applicable.

Reviewer 2: The general idea around this currently (in addition to above), is that in the absence of a profile implementation by an operator (if the cluster admin switches to that profile) none of the metrics will be exported out (the assumption is that they want to drop everything in case of `minimal`).

Author: Removed this part of this section. Furthermore since we now only support two profiles and it's out of the goal of adding further profiles I feel like this is not relevant anymore.

> - What happens if a user provides an invalid value for a matrics collection profile?

Reviewer 2: ```suggestion
- What happens if a user provides an invalid value for a metrics collection profile?
```

  - CMO will reconcile and validate that the value supplied is invalid and it
    will report Degraded=False and fail reconciliation.

- Should we add future profiles? How would we validate such profiles?
  - Our current validation strategy with only two profiles is quite linear,
    however, things start becoming more complex and hard to mainain as we
    introduce new profiles to the mix. 
  - Some of the things to consider if new profiles are introduce are:

>       - How would we validate such profile?

Reviewer 2: Profile-specific validation logic can be implemented in the [same manner as for minimal](https://github.com/rexagod/cpv/blob/main/internal/profiles/profiles.go#L10) (although the minimal one needs a bit more polishing (a better way to link SMs with rule files)).

Reviewer 2: Feel free to resolve or add the aforementioned link.

>       - How would we ensure teams that adopted metrics collection profiles implement the new profile?

Reviewer 2: [`cpv#impl-stats.go`](https://github.com/rexagod/cpv/blob/main/internal/profiles/impl-stats.go) can report partially implemented collection profiles.

Reviewer 2: Feel free to resolve or add the aforementioned link.

>       - How would we aid developers implementing the new profile?

Reviewer 2: [`cpv#minimal_guess.go`](https://github.com/rexagod/cpv/blob/main/internal/profiles/minimal_guess.go) can help them guess metrics based on a pre-defined rule file or a set of constraints they pass.

Reviewer 2: Feel free to resolve or add the aforementioned link.

### Drawbacks

- Extra CI cycles

## Design Details

### Open Questions

### Test Plan

- Unit tests in CMO to validate that the correct monitors are being selected
- E2E tests in CMO to validate that everything works correctly 
- For the `minimal` profile, origin/CI test to validate that every metric used
in a resource (Alerts/PrometheusRules/Dashboards) exist in the `keep` expression
of a minimal ServiceMonitor.

> ### Graduation Criteria

Reviewer 2: I think this needs to be updated now.

Author: What do you mean? What do you feel it's missing?

Reviewer 2: ```yaml
- Released as TechPreview: the default being `full`, it
shouldn't impact operations.
- Plan to GA: this would officially support the "minimal"
profile out-of-the-box and removes the earlier-imposed
TechPreview gate.
```
I was wondering if we can add something along these lines, to reflect the current state of the feature.

Reviewer: The GA plan is in the following section iiuc :)

Reviewer 2: My bad, I should've provided a diff for the suggestion above.

```diff
- Plan to release as TechPreview as the first step: the default being `full`, it
- shouldn't impact operations.
+ - Released as TechPreview: the default being `full`, it
+ shouldn't impact operations.
+ - Plan to GA: this would officially support the "minimal"
+ profile out-of-the-box and removes the earlier-imposed
+ TechPreview gate. PTAL below for more details.
```

    Then it would provide prometheus operator with this list.
  - Prometheus-operator with the list provided by this new component would know
    what scraping targets it could change to keep certain metrics.
- Recently Azure also added support for metrics collection profiles:
  - [Azure
    Docs](https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-scrape-configuration-minimal)
  -  https://github.com/Azure/prometheus-collector
  - In their approach they also have [hardcoded](https://github.com/Azure/prometheus-collector/blob/66ed1a5a27781d7e7e3bb1771b11f1da25ffa79c/otelcollector/configmapparser/tomlparser-default-targets-metrics-keep-list.rb#L28)
  set of metrics that are only consumed when the minimal profile is enabled.
  However, customer are also able to extend this minimal profile with regexes to
  include metrics which might be interesting to them.

## Infrastructure Needed [optional]

### Adopted metrics collection profiles

Add the team and the component that will to adopt metrics collection profiles
and implementation status. Possible implementation status: 
- considering

> - implementation in progress

Reviewer 2: This should be able to tell us _for which profile is the implementation is in progress_ as well (to distinguish between multiple profiles in the future).

Author: Since we now only support two profiles and it's out of the goal of adding further profiles I feel like this is not relevant anymore.

---
title: metrics-collection-profiles
authors:
  - @JoaoBraveCoding
reviewers:
  - @openshift/openshift-team-monitoring
approvers:
  - TBD
api-approvers: "None"
creation-date: 2022-12-06
last-updated: 2023-07-24
tracking-link:
  - https://issues.redhat.com/browse/MON-2483
  - https://issues.redhat.com/browse/MON-3043

> ---

Reviewer 4: from doc view, https://github.com/openshift/enhancements/blob/7a07c6ff5aca6857c3dd8669d6f58054bc64b348/enhancements/monitoring/metrics-collection-profiles.md
![yaml](https://github.com/juzhao/test/assets/22743651/675f4a36-89fb-420f-8983-89e719a666e4)
should be below
```
authors:
  - "@JoaoBraveCoding"
reviewers:
  - "@openshift/openshift-team-monitoring"
```

## Summary

The core OpenShift components ship a large number of metrics. A 4.12-nightly
cluster on AWS (3 control plane nodes + 3 worker nodes) currently produces
around 350,000 unique timeseries, and adding optional operators increases that
number. Users have repeatedly asked for a supported method of making Prometheus
consume less memory and CPU, either by increasing the scraping interval or by
scraping fewer targets.

These modifications are currently not possible, because the OpenShift operators
manage their monitors (either directly or via the cluster-version-operator) and
any manual modification would be reverted. This proposal outlines a solution for
allowing users to control the amount of data being collected which can meet
their needs.

## Motivation

OpenShift is an opinionated platform, and the default set of metrics has of
course been crafted to match what we think the majority of users might need.
Nevertheless, users have repeatedly asked for the ability to reduce the amount
of memory consumed by Prometheus either by lowering the Prometheus scrape
intervals or by modifying monitors.

> Users currently can not control the aformentioned monitors scraped by Prometheus

Reviewer 4: ```suggestion
Users currently can not control the aforementioned monitors scraped by Prometheus
```

cluster. The analysis consisted of a test where an OpenShift cluster would run
two replicas of the OpenShift Prometheus instance but each replica would be
configured to a different metrics collection profile (`full`, `minimal`). Then
we would trigger a workload using
[kube-burner](https://github.com/cloud-bulldozer/kube-burner) and at the end of
2 hours, we evaluated the results. We concluded that in terms of resource usage,
given the results obtained we can confidently state that the feature is quite
valuable given the reduction of CPU by ~21% and memory by ~33% when comparing
the `minimal` to the `full` profile. More details can be consulted in this
[document](https://docs.google.com/document/d/1MA-HTJQ_X7y_bwpJS2IPbGmC4qDyIMp25jisr34X2F4/edit?usp=sharing).

Moreover, through Telemetry, we collect for each cluster the top 3 Prometheus
jobs that generate the biggest amount of samples. With this data, we know that
for OpenShift 4.11 the 5 components most often reported as the biggest producers
are: the Kubernetes API servers, the Kubernetes schedulers, kube-state-metrics,
kubelet and the network daemon.

### User Stories

> - As an user, I want to lower the amount of resources consumed by Prometheus in

Reviewer 4: ```suggestion
- As a user, I want to lower the amount of resources consumed by Prometheus in
```

want to be included in a given profile.
- Support the addition of profiles other than `full` and `minimal`.

## Proposal

This enhancement leverages label selectors to allow cluster-monitoring-operator
to select the appropriate monitors for a given metrics collection profile. 

Given this, the following steps would be necessary to implement this
enhancement:

- expose a configuration option in the cluster-monitoring-operator ConfigMap
  that allows selecting the metrics collection profile;
- OCP components that would like to implement metrics collection profiles would
  have to provide at least 1 monitor for each supported metrics collection
  profile.

### Workflow Description

The cluster-monitoring-operator (CMO) will be extended with a new field,
`collectionProfile` under `prometheusK8s`. This new field will then allow users
to configure their desired metrics collection profile while preserving the
platform functionality (alerts, console, HPA and VPA and telemetry).

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:

>       collectionProfile: full 

Reviewer 4: remove the space after `full`
```suggestion
      collectionProfile: full
```

ServiceMonitor per profile).

```yaml
---
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    k8s-app: telemeter-client
    monitoring.openshift.io/collection-profile: full
  name: telemeter-client
  namespace: openshift-monitoring
spec:
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    interval: 30s
    port: https
    scheme: https
    tlsConfig:
      <...>
  jobLabel: k8s-app
  selector:
    matchLabels:
      k8s-app: telemeter-client
---
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    k8s-app: telemeter-client
    monitoring.openshift.io/collection-profile: minimal
  name: telemeter-client-minimal
  namespace: openshift-monitoring
spec:
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    interval: 30s
    port: https
    scheme: https
    tlsConfig:
      <...>
  jobLabel: k8s-app
  selector:
    matchLabels:
      k8s-app: telemeter-client
  metricRelabelings:
  - sourceLabels: [__name__]
    action: keep
    regex: "federate_samples|federate_filtered_samples"
 ```

Note: 

> - the `metricRelabeling`s section keeps only two metrics, while the rest is

Reviewer 4: ```suggestion
- the `metricRelabelings` section keeps only two metrics, while the rest is
```

dropped.
- the metrics in the `keep` section were obtained with the help of a script that
  parsed all Alerts, PrometheusRules and Console dashboards to determine what
  metrics were actually being used.

Finally, a team that adopts the metrics collection profile feature should also 
add themselves and their component to the list under 

> [Infrastruture needed](#infrastructure-needed-optional)

Reviewer 4: ```suggestion
[Infrastructure needed](#infrastructure-needed-optional)
```

	// profile, Prometheus only collects metrics necessary for the default
	// platform alerts, recording rules, telemetry and console dashboards.
	CollectionProfile CollectionProfile `json:"collectionProfile,omitempty"`
}
```

### Implementation Details/Notes/Constraints

Each OpenShift team that wants to adopt this feature will be responsible for
providing the different monitors. This work is not trivial. Dependencies between
operators and their metrics exist. This makes it difficult for developers to
determine whether a given metric can be excluded from the `minimal` profile or
not. To aid teams with this effort the monitoring team will provide:
- a CLI tool that consumes a ServiceMonitor and a list of metrics and
  generates the `minimal` ServiceMonitor for that operator.
- an origin/CI test that validates for all Alerts and PrometheusRules that the
  metrics used by them are present in the `keep` expression of the
  monitor for the `minimal` profile

### Risks and Mitigations

- How are monitors supposed to be kept up to date? In 4.12 a metric that wasn't
  being used in an alert is now required, how does the monitor responsible for

>   that metric gets updated?

Reviewer 4: ```suggestion
  that metric get updated?
```

  - The origin/CI test mentioned in the previous section will fail if there is a
    resource (Alerts/PrometheusRules/Dashboards) using a metric which is not
    present in the monitor in question;

- What happens if a user provides an invalid value for a metrics collection profile?
  - CMO will reconcile and validate that the value supplied is invalid and it
    will report Degraded=False and fail reconciliation.

- Should we add future profiles? How would we validate such profiles?
  - Our current validation strategy with only two profiles is quite linear,

>     however, things start becoming more complex and hard to mainain as we

Reviewer 4: ```suggestion
    however, things start becoming more complex and hard to maintain as we
```

    introduce new profiles to the mix. 
  - Some of the things to consider if new profiles are introduce are:
      - How would we validate such profile?
      - How would we ensure teams that adopted metrics collection profiles
        implement the new profile?
      - How would we aid developers implementing the new profile?
### Drawbacks

- Extra CI cycles

## Design Details

### Open Questions

### Test Plan

- Unit tests in CMO to validate that the correct monitors are being selected
- E2E tests in CMO to validate that everything works correctly 
- For the `minimal` profile, origin/CI test to validate that every metric used

> in a resource (Alerts/PrometheusRules/Dashboards) exist in the `keep` expression

Reviewer 4: ```suggestion
in a resource (Alerts/PrometheusRules/Dashboards) exists in the `keep` expression
```

of a minimal monitors.

### Graduation Criteria

- Released as TechPreview: the default being `full`, it
shouldn't impact operations.

- Plan to GA: this would officially support the "minimal"
profile out-of-the-box and removes the earlier-imposed
TechPreview gate. PTAL at the section below for more details.

#### Tech Preview -> GA

- [Automation to update metrics in collection profiles](https://issues.redhat.com/browse/MON-3106)
- [Telemetry signal for collection profile usage](https://issues.redhat.com/browse/MON-3231)
- [Enhancement proposal for metrics collection profile](https://issues.redhat.com/browse/MON-2692)
- [CLI tool to facilitate implementation of metrics collection profiles for OCP components](https://issues.redhat.com/browse/MON-2694)
- [Drop TechPreview gate on collection profiles](https://issues.redhat.com/browse/MON-3215)
- [origin/CI tool to validate collection profiles](https://issues.redhat.com/browse/MON-3105)

> - [User facing documentation created in openshift-docs](https://issues.redhat.com/browse/OBSDOCS-330)

Reviewer 4: ```suggestion
- [User facing documentation created in OpenShift-docs](https://issues.redhat.com/browse/OBSDOCS-330)
```

  - CMO acting behind the back of other operators isn't probably a very good
    idea (less explicit, more brittle);
  - it's likely to be less efficient because the relabeling configs would be
    very complex and expensive in terms of processing.
  - Hypershift implements a [similar
    strategy](https://github.com/openshift/hypershift/pull/1294), but again this
    would only work for service monitors that CMO deploys directly
- Let users configure themselves the Prometheus scraping interval. This solution
  was discarded because:
  - Users might not be fully aware of the impact that changing this interval
    might have. Too low of an interval and users might overwhelm Prometheus and
    exhaust its memory. In contrast, too long of an interval might render the
    default alerts ineffective;
  - It's not advisable to increase the scrape interval to more than 2 minutes.
    Staleness occurs at 5 minutes (by default) which is would cause gaps in
    graphs/dashboards;
  - Many upstream dashboards were built while having the 30 second interval in
    mind;
  - Scrapes can fail, but the user might not be mindful of this and set a high
    scrape interval;

> - Add a seperate container to prometheus-operator (p-o) that would be used by

Reviewer 4: ```suggestion
- Add a separate container to prometheus-operator (p-o) that would be used by
```

  p-o to modify the prometheus config according to a metrics collection profile.
  - This container would perform an analysis on what metrics were being used.
    Then it would provide prometheus operator with this list.
  - Prometheus-operator with the list provided by this new component would know
    what scraping targets it could change to keep certain metrics.
- Recently Azure also added support for metrics collection profiles:
  - [Azure
    Docs](https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-scrape-configuration-minimal)
  -  https://github.com/Azure/prometheus-collector
  - In their approach they also have [hardcoded](https://github.com/Azure/prometheus-collector/blob/66ed1a5a27781d7e7e3bb1771b11f1da25ffa79c/otelcollector/configmapparser/tomlparser-default-targets-metrics-keep-list.rb#L28)
  set of metrics that are only consumed when the minimal profile is enabled.

>   However, customer are also able to extend this minimal profile with regexes to

Reviewer 4: ```suggestion
  However, customers are also able to extend this minimal profile with regexes to
```

  include metrics which might be interesting to them.

> 

Author: Add reference to https://docs.google.com/document/d/1I-YT7LKKDHSBLB6Hmg0tZ54DWjrAxlVdXxlViShMu-0/edit

## Infrastructure Needed [optional]

### Adopted metrics collection profiles

> Add the team and the component that will to adopt metrics collection profiles

Reviewer 4: ```suggestion
Add the team and the component that will adopt metrics collection profiles
```

[document](https://docs.google.com/document/d/1MA-HTJQ_X7y_bwpJS2IPbGmC4qDyIMp25jisr34X2F4/edit?usp=sharing).

Moreover, through Telemetry, we collect for each cluster the top 3 Prometheus
jobs that generate the biggest amount of samples. With this data, we know that
for OpenShift 4.11 the 5 components most often reported as the biggest producers
are: the Kubernetes API servers, the Kubernetes schedulers, kube-state-metrics,
kubelet and the network daemon.

### User Stories

- As an user, I want to lower the amount of resources consumed by Prometheus in
  a supported way, so I can configure the clusters metrics collection profiles
  to `minimal`.
- As a developer, I want a supported way to collect a subset of the metrics
  exported by my operator and operands.

### Goals

- Give users a supported and documented method to pick the amount of metrics
  being collected by the platform, in a way that allows
  cluster-monitoring-operator and other operators to provide their own
  defaulting.
- Other components of the cluster (alerts, console, HPA and VPA and telemetry)
  should not be negatively affected by this feature.

> - Developers may optionally adhere to this feature, those that do not adhere

Author: I didn't fix this one because it's clarified in section `Terms` that when I write `Developers` I mean OpenShift developers

  generates the `minimal` ServiceMonitor for that operator.
- an origin/CI test that validates for all Alerts and PrometheusRules that the
  metrics used by them are present in the `keep` expression of the
  monitor for the `minimal` profile

### Risks and Mitigations

- How are monitors supposed to be kept up to date? In 4.12 a metric that wasn't
  being used in an alert is now required, how does the monitor responsible for
  that metric gets updated?
  - The origin/CI test mentioned in the previous section will fail if there is a
    resource (Alerts/PrometheusRules/Dashboards) using a metric which is not
    present in the monitor in question;

- What happens if a user provides an invalid value for a metrics collection profile?
  - CMO will reconcile and validate that the value supplied is invalid and it
    will report Degraded=False and fail reconciliation.

- Should we add future profiles? How would we validate such profiles?
  - Our current validation strategy with only two profiles is quite linear,
    however, things start becoming more complex and hard to mainain as we
    introduce new profiles to the mix. 

>   - Some of the things to consider if new profiles are introduce are:

Reviewer 3: Could it work if we only introduce/support the new profile for techpreview clusters and GA when all components have been modified to support it?

Author: From my POV that would certainly solve the problem 

> How would we ensure teams that adopted metrics collection profiles
        implement the new profile?

But it's still unclear how could we validate the new profile. Regardless I can add that idea to this section

- Deprecate the feature

### Upgrade / Downgrade Strategy

- If metrics collection profiles is accepted and is released to 4.13 then we
  must backport the new monitors selectors to 4.12. The reason being that when
  an upgrade occurs and the new monitors are deployed, we will not want the 4.12
  Prometheus-Operator to process all the new monitors as this would configure
  the Prometheus instances to start double scraping the OpenShift components
  that implemented the metrics collection profiles feature. Done in
  [cluster-monitoring-operator#2047](https://github.com/openshift/cluster-monitoring-operator/pull/2047)
- Once we backport the new monitors selectors upgrades and
  downgrades are not expected to present a significant challenge.

### Version Skew Strategy

TBD but I don't think it applies here

### minimal Aspects of API Extensions

TBD but I don't think it applies here

#### Failure Modes

TBD but I don't think it applies here

#### Support Procedures

TBD but I don't think it applies here

## Implementation History

Initial proofs-of-concept:

- https://github.com/openshift/cluster-monitoring-operator/pull/1785

## Alternatives

> - Make CMO injecting metric relabelling for all service monitors based on the

Reviewer 3: I'm not even sure how it can work in practice.

Author: I had put this here just to show the reviewers that it had been something we had considered. But if you feel it doesn't make sense to have here I'm fine with removing it

Reviewer 3: Right, I'm ok keeping it here but maybe highlight how cumbersome and error-prone it would be. The only way I could see it working would be that CMO watches the component monitors and generates new resources that would be reconciled by prometheus-operator instead of the original ones.

- Recently Azure also added support for metrics collection profiles:
  - [Azure
    Docs](https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-scrape-configuration-minimal)
  -  https://github.com/Azure/prometheus-collector
  - In their approach they also have [hardcoded](https://github.com/Azure/prometheus-collector/blob/66ed1a5a27781d7e7e3bb1771b11f1da25ffa79c/otelcollector/configmapparser/tomlparser-default-targets-metrics-keep-list.rb#L28)
  set of metrics that are only consumed when the minimal profile is enabled.
  However, customer are also able to extend this minimal profile with regexes to
  include metrics which might be interesting to them.

## Infrastructure Needed [optional]

### Adopted metrics collection profiles

Add the team and the component that will to adopt metrics collection profiles
and implementation status. Possible implementation status: 
- considering
- implementation in progress
- implemented

| Team            | Component          | Implementation Status      |
|-----------------|--------------------|----------------------------|
| Monitoring Team | kubelet            | Implemented                |

> | Monitoring Team | etcd               | Implemented                |

Reviewer 3: (nit) @machine424 to confirm :)
```suggestion
| etcd Team | etcd               | Implemented                |
```

- As a developer of a component (that does not, or only partially implements a
  profile), I want to get information about any monitors that are not yet
  implemented for any of the supported profiles that are offered.
- As a developer of a component (that implements a profile), I want to verify if
  all the profile metrics are present in the cluster, and which of the profile
  monitors are affected if not. Also, I want additional information to narrow
  down where these metrics are exactly being used.

### Goals

- Give users a supported and documented method to pick the amount of metrics
  being collected by the platform, in a way that allows
  cluster-monitoring-operator and other operators to provide their own
  defaulting.
- Other components of the cluster (alerts, console, HPA and VPA and telemetry)
  should not be negatively affected by this feature.
- Developers may optionally adhere to this feature, those that do not adhere
  should not be impacted.

### Non-Goals

- Dynamically adjusting metrics scraped at runtime based on heuristics.
- Allowing users to explicitly set a list or a regex of metrics that they would
want to be included in a given profile.

> - Support the addition of profiles other than `full` and `minimal`.

Reviewer 2: Assuming the Collection Profiles operate in a decoupled manner w.r.t. installer, (1) is something we have anticipated, and have support for within the current framework. The workflow would go about like so:

- CMO introduces a `custom` profile.
- Teams implement said profile on their end, [like so](assets/kube-state-metrics/minimal-service-monitor.yaml).
- Teams configure their respective component to detect the currently configured profile and deploy the appropriate resources needed to sustain that profile.

_I believe the EP is outdated to reflect certain advancements we have made recently. I make sure these are reflected here soon, since we now have answers for [initially proposed questions](https://github.com/openshift/enhancements/pull/1298/files#diff-d53514a51b0340df4d66642f8263b41c912d81fdb8f87522fefd4f318a9e3511R341-R344)._

(2) is something that's not addressed in the current implementation, but definitely something that's good to have. While we are indeed leaning towards keeping things simple from both sides and thus onboard and [help as many teams leverage everything the featureset currently has to offer](https://github.com/rexagod/cpv), I'll bring this up with the team and update the EP with planning to implement this in the later iterations.

  - Prometheus-operator with the list provided by this new component would know
    what scraping targets it could change to keep certain metrics.
- Recently Azure also added support for metrics collection profiles:
  - [Azure
    Docs](https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-scrape-configuration-minimal)
  -  https://github.com/Azure/prometheus-collector
  - In their approach they also have
  [hardcoded](https://github.com/Azure/prometheus-collector/blob/66ed1a5a27781d7e7e3bb1771b11f1da25ffa79c/otelcollector/configmapparser/tomlparser-default-targets-metrics-keep-list.rb#L28)
  set of metrics that are only consumed when the minimal profile is enabled.
  However, customers are also able to extend this minimal profile with regexes
  to include metrics which might be interesting to them.
- Leverage [installer capabilities](https://docs.google.com/document/d/1I-YT7LKKDHSBLB6Hmg0tZ54DWjrAxlVdXxlViShMu-0/edit#heading=h.848jsje80fru)
  - After some consideration we decided to abandon this idea since it would only
    work for resources controlled by CVO which is not the case for the majority

>     of ServiceMonitors.

Reviewer 5: We, components owners already rely on installer capabilities for enabling/disabling Console related SM/Rules (e.g https://github.com/openshift/cluster-kube-apiserver-operator/pull/1542/files), so having yet another thing to configure is not very ideal considering that we will still have to maintain our code relying on the Console capability from the installer.

Would it be perhaps possible to unify these approach to make it easier to integrate and maintain for component owners? An option could be to require component owners to let their Prometheus resources be managed by CVO and then leverage installer capabilities.

@bparees could you have a look to this enhancement since you worked on installer capabilities?

Reviewer 5: Interesting. I looked into our code again as well and yes it turns out that there are some rules that are managed by our operator and not CVO and we already have some in-house conditional logic for SNO: https://github.com/openshift/cluster-kube-apiserver-operator/blob/03c5a03d1bd0a952655aa180bd2708802e8cc1a4/pkg/operator/starter.go#L216

I am just afraid that there might be too many ways at the moment to customize the data we expose and that it will be a pain to maintain going forward. Unifying all the existing knobs for monitoring under collection profiles might also be a good way to simplify the existing processes.

Reviewer 6: in addition to what @simonpasquier said, i think caps are not a good fit for this because it's easy to envision someone wanting a cap (like image-registry) but then wanting to choose the level of metrics that cap should provide (minimal vs full).  There might be ways to solve this with judicious use of cap-names on ServiceMonitor resources and the introduction of new caps, but i think it would be pretty ugly
